### ISSUE TITLE
All `GM_xmlhttpRequest` requests are serialized

### ISSUE THREAD

**Issue by derjanb** [Tampermonkey/tampermonkey#2215]
## Expected Behavior
`GM_xmlhttpRequest` requests are made in parallel

## Actual Behavior
requests run serialized

## Specifications
- Chromium: any
- TM: 5.3+

## 

In order to fix #2179 Tampermonkey needs to register a [DNR](https://developer.chrome.com/docs/extensions/reference/api/declarativeNetRequest) rule for every request. Since those header modifications are not preserved over HTTP redirects like headers set via `fetch`, Tampermonkey needs to match **all** requests, but therefore can't run more than one at once. In my opinion this is a [Chrome bug](https://github.com/w3c/webextensions/issues/694).

## Script

```js
// ==UserScript==
// @name         Non-parallel GM_xhr
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       You
// @match        https://example.com/*
// @grant        GM_xmlhttpRequest
// @grant        GM.xmlHttpRequest
// @connect      httpbin.org
// ==/UserScript==

console.log('Dummy request started');

GM.xmlHttpRequest({
    method:   "GET",
    url:      'https://httpbin.org/delay/5000',
})
.then(() => {
    console.log('Dummy request finished');
})

GM_xmlhttpRequest({
    method: 'GET',
    url: 'https://httpbin.org/redirect-to?url=https://httpbin.org/headers',
    responseType: 'arraybuffer',
    headers: {
        'user-agent': 'TM',
        'foo': 'bar'
    },
    onreadystatechange: function(event) {
        console.log(`Ready state: ${event.readyState}, status ${event.status}`);
    },
    onprogress: function(event) {
        if (event.lengthComputable) {
            console.log(`Downloading: ${event.loaded}/${event.total}`);
        }
    },
    onload: function(res) {
        console.log('Download finish');
        console.log(res);
        console.log(res.responseText);
    },
    onerror: function(res) {
        console.log('Can\'t load URL');
    }
});
```

**Comment by Purfview** [comment]
I got lots of reports about this from Chrome users, and one was Chromium user.

**Comment by derjanb** [comment]
> I got lots of reports about this from Chrome users, and one was Chromium user.

@Purfview Have you tried the workaround with the latest BETA version 5.3.6216?

If you wan't to test it at the stable version, then you have to download this [crx](https://tampermonkey.net//crx/tampermonkey_stable.crx), because version 5.3.2 is currently in review.

After downloading the crx file linked above you need to drag and drop it to the extensions page `chrome://extensions` (after you've enabled 'Developer Mode').

**Comment by Purfview** [comment]
In BETA version 5.3.6216: responses comes very slow - still serialized, and same like the other issue with loading, refreshed it like 50 times till it worked.

**Comment by derjanb** [comment]
> In BETA version 5.3.6216: responses comes very slow

The workaround from above should fix this one:
```
// @require      https://raw.githubusercontent.com/Tampermonkey/utils/refs/heads/main/requires/gh_2215_make_GM_xhr_more_parallel_again.js
```

**Comment by xmcp** [comment]
This change breaks userscripts sending multiple network requests and setting a short timeout for each request. [E.g., hanFengSan/eHunter:](https://github.com/hanFengSan/eHunter/blob/3ea92b674cd81d5b8648ca9e2fb641c1f77f7b31/src/platform/base/request/TextReq.ts#L74)

```javascript
let timeout = new Promise((resolve, reject) => {
    setTimeout(reject, this.timeoutTime * 1000 * this.curRetryTimes, 'request timed out');
});
let req = new Promise((resolve, reject) => {
    GM_xmlhttpRequest(...);
});
Promise.race([timeout, req]);
```

Since `GM_xmlhttpRequest` now send requests one by one, one slow response will block all following requests and make them timed-out. Therefore such scripts need the `@require` workaround to mitigate this problem.

Ocrosoft/PixivPreviewer, which referenced this issue above, is likely also affected.

Since this problem is ultimately caused by a missing feature in Chrome API, can we expect this sequential behavior will exist for a long time? If so, should each affected userscript add the workaround themselves, or Tampermonkey can somehow integrate the workaround to avoid breaking these scripts?

**Comment by derjanb** [comment]
> Since this problem is ultimately caused by a missing feature in Chrome API, can we expect this sequential behavior will exist for a long time?

This depends on w3c/webextensions/issues/694

> Tampermonkey can somehow integrate the workaround to avoid breaking these scripts

As mentioned above, the workaround has drawbacks which userscripts can't automatically handle. So automatically adding it, would break scripts that use `GM_xhr` with URLs that might see a redirect for authentication. Also the workaround is a hack and therefore should not be used automatically.
That's why userscript developers need to manually `@require` the workaround and then can make sure that the userscript is prepared for side-effects.
On the other hand, the timing of network requests is never guaranteed, so I expect more flexibility here without breaking things. Of course, userscripts that depend on a certain timing can break.

> Is the serialization global

Unfortunately, yes! [It's even extension wide - affecting internal requests as well](https://github.com/w3c/webextensions/issues/694#issuecomment-2377623605).
But the global lock is only active until the request headers are received. Tampermonkey doesn't have to wait for the end of the request.

> Is adding the workaround only to userscripts I maintain enough to opt-out serialization [...] will it still block requests from my userscript even if I used the workaround?

No. Every running userscript that uses `GM_xhr` without `redirect: 'manual'` can activate the lock by making a request.

Feel free to add a (constructive 😉) post to w3c/webextensions/issues/694.

**Comment by theowenyoung** [comment]
I've found that the workaround with `gh_2215_make_GM_xhr_more_parallel_again.js` works for requests to different URLs, but requests to the same URL are still processed serially.  

Version: v5.3.3

I created a minimal test script to verify this behavior:

```javascript
// ==UserScript==
// @name         Non-parallel GM_xhr
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       You
// @match        https://example.immersivetranslate.com/*
// @grant        GM_xmlhttpRequest
// @grant        GM.xmlHttpRequest
// @connect      httpbin.org
// @require      https://github.com/Tampermonkey/utils/raw/d8a4543a5f828dfa8eefb0a3360859b6fe9c3c34/requires/gh_2215_make_GM_xhr_more_parallel_again.js
// ==/UserScript==


GM.xmlHttpRequest({
    url: 'https://httpbin.org/delay/1000',
    onreadystatechange: function(event) {
        console.log(`Ready 1 state: ${event.readyState}, status ${event.status}`);
    },
})
.then(() => {
    console.log('Dummy 1 request finished');
})
GM.xmlHttpRequest({
    url: 'https://httpbin.org/delay/1000',
    onreadystatechange: function(event) {
        console.log(`Ready 2 state: ${event.readyState}, status ${event.status}`);
    },
})
.then(() => {
    console.log('Dummy 2 request finished');
})
```

When both requests target the same URL, they still run serially (taking ~2 seconds total), with the second request only starting after the first completes. Is this an intended limitation of the workaround, or is there a way to make identical-URL requests run in parallel as well?

This is important for my use case where I need to make multiple POST requests to the same endpoint.

**Comment by derjanb** [comment]
> I've found that the workaround with gh_2215_make_GM_xhr_more_parallel_again.js works for requests to different URLs, but requests to the same URL are still processed serially.

Yes. This will for sure be fixed by https://github.com/w3c/webextensions/issues/786 and I'll try another potential workaround that was mentioned by @Rob--W today.
